cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpecGenerator.java

```diff
@@ -136,41 +138,54 @@ protected void configureStages(ETLConfig config, PipelineSpec.Builder specBuilde
-        // no need to set any input schemas for an Action plug
-        if (Action.PLUGIN_TYPE.equals(outputStageType)) {
-          continue;
-        }
+        String nextStageType = pluginTypes.get(nextStageName);
 
-        DefaultStageConfigurer outputStageConfigurer = pluginConfigurers.get(outputStageName).getStageConfigurer();
+        DefaultStageConfigurer outputStageConfigurer = pluginConfigurers.get(nextStageName).getStageConfigurer();
 
         // Do not allow null input schema for Joiner
-        if (BatchJoiner.PLUGIN_TYPE.equals(outputStageType) && outputSchema == null) {
+        if (BatchJoiner.PLUGIN_TYPE.equals(nextStageType) && outputSchema == null) {
           throw new IllegalArgumentException(String.format("Joiner cannot have any null input schemas, but stage %s " +
                                                              "outputs a null schema.", stageName));
         }
 
         // if the output stage is an error transform, it takes the error schema of this stage as its input.
+        // if the current stage is a splitter transform, it takes the output schema of the port it is connected to
         // all other plugin types that the output schema of this stage as its input.
-        Schema nextStageInputSchema = ErrorTransform.PLUGIN_TYPE.equals(outputStageType) ?
-          outputErrorSchema : outputSchema;
+        Schema nextStageInputSchema;
+        if (ErrorTransform.PLUGIN_TYPE.equals(nextStageType)) {
+          nextStageInputSchema = configuredStage.stageSpec.getErrorSchema();
+        } else if (SplitterTransform.PLUGIN_TYPE.equals(configuredStage.stageSpec.getPlugin().getType())) {
+          StageSpec.Port portSpec = configuredStage.stageSpec.getOutputPorts().get(nextStageName);
+          // port can be null if no output ports were specified at configure time
+          // this can happen if the ports are dependent on the data received by the plugin
+          if (portSpec == null) {
+            nextStageInputSchema = null;
+          } else if (portSpec.getPort() == null) {
+            // null if a splitter was connected to another stage without a port specified.
+            // Should not happen since it should have been validated earlier, but check here just in case
+            throw new IllegalArgumentException(
+              String.format("Must specify a port when connecting Splitter '%s' to '%s'", stageName, nextStageName));
+          } else {
+            nextStageInputSchema = portSpec.getSchema();
+          }
+        } else {
+          nextStageInputSchema = configuredStage.stageSpec.getOutputSchema();
+        }
 
-        // Do not allow more than one input schema for stages other than Joiner
-        if (!BatchJoiner.PLUGIN_TYPE.equals(outputStageType) &&
+        // Do not allow more than one input schema for stages other than Joiner and Action
+        if (!BatchJoiner.PLUGIN_TYPE.equals(nextStageType) && !Action.PLUGIN_TYPE.equals(nextStageType) &&
```

Comment:
i changed something a little higher so that the actions would store this information in case we need it in the future, and to consolidate some logic.